pkg/resources/mutation/cluster/provisioningcluster.go

```diff
@@ -53,10 +100,250 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
+func GetKubeAPIServerArg(cluster *v1.Cluster) map[string]string {
+	if cluster.Spec.RKEConfig.MachineGlobalConfig.Data != nil {
+		return toMap(cluster.Spec.RKEConfig.MachineGlobalConfig.Data["kube-apiserver-arg"])
+	}
+	return map[string]string{}
+}
+
+func toMap(input interface{}) map[string]string {
+	args := map[string]string{}
+	parsed := convert.ToInterfaceSlice(input)
+	for _, arg := range parsed {
+		key, val, found := strings.Cut(convert.ToString(arg), "=")
+		if !found {
+			logrus.Debugf("skipping argument [%s] which does not have right format", arg)
+			continue
+		}
+		args[key] = val
+	}
+	return args
+}
+
+// setKubeAPIServerArg uses the provided arg to overwrite the value of kube-apiserver-arg under the cluster's MachineGlobalConfig.
+// If the provided arg is an empty map, setKubeAPIServerArg removes the existing kube-apiserver-arg from the cluster's MachineGlobalConfig.
+func setKubeAPIServerArg(arg map[string]string, cluster *v1.Cluster) {
+	if len(arg) == 0 {
+		delete(cluster.Spec.RKEConfig.MachineGlobalConfig.Data, "kube-apiserver-arg")
+		return
+	}
+	parsed := make([]any, 0, len(arg))
+	for key, val := range arg {
+		parsed = append(parsed, fmt.Sprintf("%s=%s", key, val))
+	}
+	if cluster.Spec.RKEConfig.MachineGlobalConfig.Data == nil {
+		cluster.Spec.RKEConfig.MachineGlobalConfig.Data = make(map[string]interface{})
+	}
+	cluster.Spec.RKEConfig.MachineGlobalConfig.Data["kube-apiserver-arg"] = parsed
+}
+
+// MachineSelectorFileForPSA generates an RKEProvisioningFiles that mounts the secret which contains
+// the generated admission configuration file to the control plane node
+func MachineSelectorFileForPSA(secretName, mountPath string) rkev1.RKEProvisioningFiles {
+	return rkev1.RKEProvisioningFiles{
+		MachineLabelSelector: &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				controlPlaneRoleLabel: "true",
+			},
+		},
+		FileSources: []rkev1.ProvisioningFileSource{
+			{
+				Secret: rkev1.K8sObjectFileSource{
+					Name: secretName,
+					Items: []rkev1.KeyToPath{
+						{
+							Key:  secretKey,
+							Path: mountPath,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// addMachineSelectorFile adds the provided RKEProvisioningFiles to the cluster's MachineSelectorFiles list.
+// It is a no-op if the provided RKEProvisioningFiles already exist in the cluster.
+func addMachineSelectorFile(file rkev1.RKEProvisioningFiles, cluster *v1.Cluster) {
+	for _, item := range cluster.Spec.RKEConfig.MachineSelectorFiles {
+		if equality.Semantic.DeepEqual(file, item) {
+			return
+		}
+	}
+	cluster.Spec.RKEConfig.MachineSelectorFiles = append(cluster.Spec.RKEConfig.MachineSelectorFiles, file)
+}
+
+// dropMachineSelectorFile removes the provided RKEProvisioningFiles from the cluster if it is found in the cluster.
+func dropMachineSelectorFile(file rkev1.RKEProvisioningFiles, cluster *v1.Cluster) {
+	source := cluster.Spec.RKEConfig.MachineSelectorFiles
+	for i := len(source) - 1; i >= 0; i-- {
```

Comment:
Leave a comment like before that we are doing this for faster lookups.

Reply:
fixed.